File: plot/src/main/java/net/fwitz/math/plot/bifurc/LogisticMapRenderer.java
package net.fwitz.math.plot.bifurc;

import net.fwitz.math.numth.numbers.Randomizer;
import net.fwitz.math.plot.ImageRenderer;

import java.awt.*;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicReferenceArray;
import java.util.stream.DoubleStream;

import static java.util.Objects.requireNonNull;

public class LogisticMapRenderer extends ImageRenderer {
    private final double[] rValues;
    private final int[] blackColumn;
    private final double minxn;
    private final double maxxn;
    private final double xnScale;

    private final AtomicReferenceArray<double[]> values;
    private final LogisticMapPanel panel;

    public LogisticMapRenderer(LogisticMapPanel panel, int width, int height) {
        super(width, height);

        this.panel = requireNonNull(panel, "panel");
        this.rValues = partition(width, panel.minr, panel.maxr);
        if (panel.minxn > panel.maxxn) {
            this.minxn = panel.maxxn;
            this.maxxn = panel.minxn;
        } else {
            this.minxn = panel.minxn;
            this.maxxn = panel.maxxn;
        }
        this.xnScale = height / (this.maxxn - this.minxn);
        this.values = new AtomicReferenceArray<>(width);
        this.blackColumn = new int[height];

        Arrays.fill(blackColumn, Color.BLACK.getRGB());
    }

    public void render() {
        pipeline.flush();
        Arrays.stream(Randomizer.shuffledInts(width))
                .mapToObj(this::colRenderer)
                .forEach(pipeline::execute);
    }

    private Runnable colRenderer(int x) {
        return () -> {
            if (!pipeline.isShutdown()) {
                calculateColWithExceptionsHandled(x);
            }
        };
    }

    private void calculateColWithExceptionsHandled(int x) {
        try {
            final double[] values = calculateColValues(x);
            if (pipeline.isShutdown()) {
                return;
            }

            this.values.set(x, values);
            if (pipeline.isShutdown()) {
                return;
            }

            paintCol(x, values);
            if (pipeline.isShutdown()) {
                return;
            }

            panel.repaint();
        } catch (Exception e) {
            System.err.println("Unexpected exception while rendering x=" + x + "; r=" + rValues[x]);
            e.printStackTrace();
        }
    }

    private double[] calculateColValues(int x) {
        return panel.computeFn.apply(rValues[x]);
    }

    protected void paintCol(int x, double[] values) {
        if (values == null) {
            return;
        }

        image.setRGB(x, 0, 1, height, blackColumn, 0, 0);
        if (values.length == 0) {
            return;
        }

        float val = 1.0f / values.length;
        int color = new Color(1 - val, val, val / 2).getRGB();
        DoubleStream.of(values)
                .mapToInt(xn -> height - (int)Math.round((xn - minxn) * xnScale))
                .filter(y -> y >= 0 && y < height)
                .forEach(y -> image.setRGB(x, y, color));
    }

    @Override
    public String toString() {
        return "ComplexFunctionRenderer[" + super.toString() + ']';
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/bifurc/LogisticMapRenderer.kt
package net.fwitz.math.plot.bifurc

import net.fwitz.math.numth.numbers.Randomizer
import net.fwitz.math.plot.renderer.ImageRenderer
import java.awt.Color
import java.util.*
import java.util.concurrent.atomic.AtomicReferenceArray
import java.util.stream.DoubleStream
import kotlin.math.roundToInt

class LogisticMapRenderer(
    private val panel: LogisticMapPanel,
    width: Int,
    height: Int
) : ImageRenderer(width, height) {
    private val rValues: DoubleArray = partition(width, panel.minr, panel.maxr)
    private val blackColumn: IntArray
    private var minxn = 0.0
    private var maxxn = 0.0
    private val xnScale: Double
    private val values: AtomicReferenceArray<DoubleArray?>

    override fun render() {
        pipeline.flush()
        Randomizer.shuffledInts(width).forEach { x ->
            pipeline.execute { colRenderer(x) }
        }
    }

    private fun colRenderer(x: Int) {
        if (!pipeline.isShutdown) calculateColWithExceptionsHandled(x)
    }

    private fun calculateColWithExceptionsHandled(x: Int) {
        try {
            val values = calculateColValues(x)
            if (pipeline.isShutdown) return

            this.values[x] = values
            if (pipeline.isShutdown) return

            paintCol(x, values)
            if (pipeline.isShutdown) return

            panel.repaint()
        } catch (e: Exception) {
            println("Unexpected exception while rendering xa$x=; r=${rValues[x]}")
            e.printStackTrace()
        }
    }

    private fun calculateColValues(x: Int) = panel.computeFn(rValues[x])

    protected fun paintCol(x: Int, values: DoubleArray?) {
        if (values == null) return

        image.setRGB(x, 0, 1, height, blackColumn, 0, 0)
        if (values.isEmpty()) return

        val value = 1.0f / values.size
        val color = Color(1 - value, value, value / 2).rgb
        DoubleStream.of(*values)
            .mapToInt { xn: Double -> height - ((xn - minxn) * xnScale).roundToInt() }
            .filter { y: Int -> y in 0 until height }
            .forEach { y: Int -> image.setRGB(x, y, color) }
    }

    override fun toString(): String {
        return "ComplexFunctionRenderer[" + super.toString() + ']'
    }

    init {
        if (panel.minxn > panel.maxxn) {
            minxn = panel.maxxn
            maxxn = panel.minxn
        } else {
            minxn = panel.minxn
            maxxn = panel.maxxn
        }
        xnScale = height / (maxxn - minxn)
        values = AtomicReferenceArray(width)
        blackColumn = IntArray(height)
        Arrays.fill(blackColumn, Color.BLACK.rgb)
    }
}